### ISSUE TITLE
[Solved] Error occurred. (Error code: 31)

### ISSUE THREAD

**Issue by gulbuhar** [SoftEtherVPN/SoftEtherVPN#148]
Im getting this on my 64bit gentoo box, any idea how to solve it?
I also get the same thing even with vpncmd running as root

~# vpnclient start
~$vpncmd 
 or 
~# vpncmd

> Select 1, 2 or 3: 2
> 
> Specify the host name or IP address of the computer that the destination VPN Client is operating on. 
> If nothing is input and Enter is pressed, connection will be made to localhost (this computer).
> Hostname of IP Address of Destination: 
> 
> Connected to VPN Client "localhost".
> VPN Client>remoteenable
> RemoteEnable command - Allow Remote Management of VPN Client Service
> The command completed successfully.
> 
> VPN Client>niccreate
> NicCreate command - Create New Virtual Network Adapter
> Virtual Network Adapter Name: test
> 
> Error occurred. (Error code: 31)
> Installation of the Virtual Network Adapter device driver failed.
> 
> VPN Client>

**Comment by gulbuhar** [comment]
i just find the solution
TUN/TAP needs to be enabled to create Virtual Network Adapter to do so
regenerate the kernel with 
Device Drivers  --->
  Network device support  --->
    [*] Universal TUN/TAP device driver support

**Comment by Eric7Coding** [comment]
excuse me, how do you solve this problem finally? I met this problem before, but I haven't found a solution yet.

**Comment by Palciny** [comment]
@Eric7Coding runing sudo ./vpnclient start  before ./vpncmd worked for me

**Comment by moatazelmasry2** [comment]
@Eric7Coding depends on your Linux distro, but as @gulbuhar mentioned, you might need to enable tun/tap in the kernel and recompile and install the kernel

**Comment by nerdyness** [comment]
For what it's worth, I ran into this issue using Ubuntu 16.04 LTS and solved it via...
```
sudo modprobe -v tunnel4
sudo modprobe -v ip_tunnel
sudo /usr/local/vpnclient/vpnclient stop
sudo /usr/local/vpnclient/vpnclient start
/usr/local/vpnclient/vpncmd
```

NOTE: this change is only temporary and after a reboot these kernel modules will be removed again. You can auto-load them on every reboot by adding them permanently to `/etc/modules` via:

`echo -e "tunnel4\nip_tunnel" | sudo tee -a /etc/modules`

**Comment by seanlowe** [comment]
For someone using WSL (Windows Subsystem for Linux), this does not entirely work. What could be some other workaround? Creating a virtual NIC in windows and somehow mapping it into the VPN client?

**Comment by adamryczkowski** [comment]
The Ubuntu 16.04 solution did not work for me on Ubuntu 18.04. Did anyone manage to run the `vpnclient` on Bionic?

**Comment by arno01** [comment]
> The Ubuntu 16.04 solution did not work for me on Ubuntu 18.04. Did anyone manage to run the `vpnclient` on Bionic?

Ubuntu 20.04 here, everything is working.

Make sure to use sudo, i.e. `sudo ./vpnclient start`.
(Didn't need to modprobe anything).

**Comment by fti-cneve** [comment]
Is this issue fixed in the newest kernel, such as 6.1.X?
